File: typescript-api/src/data/services/create-user.ts
import validator from 'validator'
import { User } from '../../domain/entities'
import { CheckUserRepository, CreateUserRepository } from "../repositories";
import { 
  DuplicateUserNameError,
  InvalidEmailError, 
  InvalidPasswordError,
  InvalidUserNameLenghtError
} from '../../domain/errors';

const minPasswordLength = 5
const minUserNameLength = 5

export class CreateUserService {
  constructor(
    private readonly createUserRepo: CreateUserRepository,
    private readonly checkUserRepo: CheckUserRepository
  ) { }

  async create(user: User): Promise<void> {
    if (!validator.isEmail(user.email)) {
      throw new InvalidEmailError(user.email)
    }

    if (user.password.length < minPasswordLength) {
      throw new InvalidPasswordError()
    }

    if (user.name.length < minUserNameLength) {
      throw new InvalidUserNameLenghtError(minUserNameLength)
    }

    const userExists = await this.checkUserRepo.exists(user)
    if (userExists) {
      throw new DuplicateUserNameError(user.name)
    }
 
    await this.createUserRepo.create(user)
  }
}

File: rust-api/src/data/services/create_user.rs
use crate::{
    data::repositories::CreateUserRepository,
    domain::entities::{ParseUserError, User},
    domain::errors::{CreateUserError, ErrorMessage},
};
use email_address::EmailAddress;
use std::sync::Arc;

pub async fn create_user_service(
    user: &User,
    repo: Arc<impl CreateUserRepository>,
) -> Result<(), CreateUserError> {
    let parsed_user = User::parse(user.name.clone(), user.email.clone(), user.password.clone())
        .map_err(|e| {
            eprintln!("ERROR: {:?}", e);
            match e {
                ParseUserError::InvalidPasswordLength(str) => {
                    CreateUserError::InvalidUserError(ErrorMessage { message: str })
                }
                ParseUserError::InvalidUserNameLength(str) => {
                    CreateUserError::InvalidUserError(ErrorMessage { message: str })
                }
            }
        })?;

    if !EmailAddress::is_valid(user.email.as_str()) {
        return Err(CreateUserError::InvalidEmailError(ErrorMessage {
            message: "invalid email address".to_string(),
        }));
    }

    repo.create(&parsed_user).await.map_err(|e| {
        eprintln!("ERROR: {:?}", e);

        CreateUserError::RepoError({
            ErrorMessage {
                message: "internal server error".to_string(),
            }
        })
    })?;

    Ok(())
}
